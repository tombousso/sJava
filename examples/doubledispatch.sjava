(package examples.doubledispatch)

(define-class Main () public
	((main args String[]) void public static
		(define objects (Object[] "a" "b" 1 2))

		(define v1 (Visitor1))
		(forEach o objects
			(v1:visit o)
		)

		(println)

		(define v2 (Visitor2))
		(forEach o objects
			(v2:visit o)
		)
	)
)

(define-class Visitor () abstract
	((<init>) void
		(super:<init>)
	)
	((visit s String) void abstract)
	((visit n Integer) void abstract)
	((visit o Object) void
		(doubleDispatch (this:visit _) o)
	)
)

(define-class Visitor1 (Visitor)
	((<init>) void
		(super:<init>)
	)
	((visit s String) void
		(println "Visitor1 received a String: " s)
	)
	((visit n Integer) void
		(println "Visitor1 received an Integer: " n)
	)
)

(define-class Visitor2 (Visitor)
	((<init>) void
		(super:<init>)
	)
	((visit s String) void
		(println "Visitor2 received a String: " s)
	)
	((visit n Integer) void
		(println "Visitor2 received an Integer: " n)
	)
)
