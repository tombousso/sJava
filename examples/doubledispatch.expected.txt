0 string1
1 string1
2 integer1
3 integer1

0 string2
1 string2
2 integer2
3 integer2
